### PR TITLE
[heft] Introduce initial implementation of 'copyFiles' action

### PR DESCRIPTION
**Pull request by D4N14L**
Adds a new default Heft action named `copyFiles`. This action can run in all build events _except_ clean, as there should be no need to copy any files on a clean.

An example implementation of the schema to add the event is here:
```json
{
      "actionKind": "copyFiles",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyFiles",
      "copyOperations": [
          {
             "sourceFolder": "src" // Optional, defaults to project root
             "fileExtensions": [ ".json", ".txt" ], // Optional
             "includeGlobs": [ "**/*.md" ], // Optional
             "excludeGlobs": [ "**/__test__/**/*.md" ], // Optional
             "destinationFolders": [ "dist" ], // Required
             "hardlink": false, // Optional, defaults to false
             "flatten": false // Optional, defaults to false
          }
      ]
    }
```

Some use cases for this plugin could include:
- copying checked-in assets that may not be required for Typescript build, or may not exist in the `src` directory, as `CopyStaticAssetsPlugin` is used for
- copying build assets after a build completes to some directory

Reasons for glob (which has perf implications) instead of a more targeted folder/file copy design:
- flexibility provided by the plugin can be used in various scenarios, from copying entire directories or subsets of entire directories to copying specific file extensions or specific file names
- targeted folder/file copy schema would be more verbose, possibly needing regular `heft.json` updates in order to keep up-to-date with which files need copying

### CODE REVIEW

File: apps/heft/src/plugins/CopyGlobsPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyGlobsPlugin.ts:1]: Every single time a well-meaning glob like `"src/**/*.md"` appears in one of our config files, it will most likely be processed as an inefficient `O(n)` brute force scan of the folder tree.  

- Can we call this out in the docs?  
- Can we design the tooling in a way that minimizes usage of globs?
- If not, could we invent some kind of fancy file cache that performs the expensive disk scan only once, and then matches globs against an efficient index?

File: apps/heft/src/plugins/CopyGlobsPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            resolvedTargetPathsMap = new Map<string, boolean>();
+            fileOperationMap.set(resolvedSourceFilePath, resolvedTargetPathsMap);
+          }
+
+          for (const targetFolder of copyFilesEventAction.targetFolders) {
+            resolvedTargetPathsMap.set(
+              path.resolve(
+                heftConfiguration.buildFolder,
+                targetFolder,
+                path.basename(resolvedSourceFilePath)
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyGlobsPlugin.ts:97]: Calling `path.basename()` here seems incorrect. For example, suppose that `globsToCopy` matches two file paths:

- `src/fruit/apple.md` 
- `src/phones/apple.md`

The current implementation will map them both to the path `dist/apple.md`, either silently overwriting the second one, or maybe reporting an error that the other promise is already holding a file lock.

I think most people would expect the output to be:

- `dist/fruit/apple.md` 
- `dist/phones/apple.md`

If there really is a situation where someone would want to condense a nested tree into a flat list (?), that should be an opt-in behavior.  Maybe it should even have its own action kind (`copyAndFlatten`?).

File: apps/heft/src/plugins/CopyGlobsPlugin.ts (hunk 1 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts:1]: We have the `CachedFileSystem` type I believe in Rush, however I'm not sure using that here is a fantastic plan. We could alternatively allow for specifying exact folders/file paths that need to be copied within a project. We could also do a mixed approach, where we allow globs to select files but also allow for copying directories at exact locations. This would allow us the flexibility of regex when needed and the perf of copying specific directories when we know exactly where to look.

File: apps/heft/src/plugins/CopyGlobsPlugin.ts (hunk 2 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts:97]: The issue with this design is that we would have to be able to specify what folder the path would be relative from, since the copy applies across the project, not _just_ in the `src` folder, so your example would actually produce:
- `dist/src/fruit/apple.md`
- `dist/src/phones/apple.md`

Which is why I went with flatten by default. However, if we knew what path we would want the files relative to (such as the case when the file match glob is `some/path/**/*.md`, then another property would definitely make sense for that. I'm just not sure if that wouldn't be better served by making multiple explicit action definitions, which specify specific directories, ex.
```json
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "globsToCopy": ["src/phones/**/*.md"],
      "targetFolders": ["dist/phones"],
      "hardlink": false
},
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "globsToCopy": ["src/fruits/**/*.md"],
      "targetFolders": ["dist/fruits"],
      "hardlink": false
}
```

Though... this in and of itself is pretty verbose. It seems like containing it all within one definition would be better, ie.
```json
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "copyOperations": [
          {
             "sourceGlobs": [ "src/phones/**/*.md" ],
             "targetFolders": [ "dist/phones" ],
             "hardlink": false
          },
          {
             "sourceGlobs": [ "src/fruits/**/*.md" ],
             "targetFolders": [ "dist/fruits" ],
             "hardlink": false
          }
      ]
}
```

Thoughts?

Comment (D4N14L) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts:97]: Combining this with some thoughts from @iclanton, what if we did something like this?
```JSON
{
      "actionKind": "copyFiles",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyFiles",
      "copyOperations": [
          {
             "sourceFolder": "src" // Optional, defaults to project root
             "includeGlobs": [ "**/*.md" ], // Required
             "excludeGlobs": [ "**/__test__/**/*.md" ], // Optional, defaults to []
             "targetFolders": [ "dist" ], // Required
             "hardlink": false, // Optional, defaults to false
             "flatten": false // Optional, defaults to false (true?)
          }
      ]
}
```
This solution has the ability to narrow down the scope of the glob being used, to reduce the perf impact of the search (as per your first comment). It also allows for optionally flattening/retaining the relative path from the `sourceFolder` to the matched file. Additionally, it provides flexibility to specify glob patterns to exclude from included results.

Comment (iclanton) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts:97]: I like this design. We can also optimize around some glob patterns like `**/*` and `**/*.<ext>`

File: apps/heft/src/plugins/CopyGlobsPlugin.ts (hunk 1 above)
Comment (octogonz) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts]: It would be useful for this PR to more clearly describe some real world problems that "copyFiles" is meant to solve.  Do any of them actually require globs at all? Hard to say without knowing the specifics.

File: apps/heft/src/plugins/CopyGlobsPlugin.ts (hunk 2 above)
Comment (octogonz) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts]: All these examples still have `**` in them. That is a brute force `O(n)` scan.  The way the plugin is implemented, this scan will be performed repeatedly for each `copyFiles` action. 

Planning to one day cache the scans is a good idea, but let's think about a cache for a moment:

Design 1:
```
  "staticAssetsToCopy": {
    "fileExtensions": [".css", ".jpg", ".png"]
  }
```

Design 2:
```
  "staticAssetsToCopy": {
    "includeGlobs": [
      "src/**/*.css",
      "src/**/*.jpg",
      "src/**/*.png"
    ]
  }
```

Design 1 could easily be optimized as a single scan of the `src` folder that collects files based on their file extension.

Whereas Design 2 requires invoking a complex glob engine multiple times.  The engine has to support negative lookback patterns. It has to be able to scan paths that are not under the `src` folder.  Smart engineers can write their globs efficiently, but the engine needs to support very inefficient globs.

So planning to cache later does not get us out of thinking about performance when designing the input file.

File: apps/heft/src/plugins/CopyGlobsPlugin.ts (hunk 1 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyGlobsPlugin.ts]: I've thrown some more info into the summary

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+} from '../utilities/CoreConfigFiles';
+import {
+  IBuildStageContext,
+  IBundleSubstage,
+  ICompileSubstage,
+  IPostBuildSubstage,
+  IPreCompileSubstage
+} from '../stages/BuildStage';
+
+const globEscape: (unescaped: string[]) => string[] = require('glob-escape'); // No @types/glob-escape package exists
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:29]: Can we say

```suggestion
import globEscape: (unescaped: string[]) => string[] = require('glob-escape'); // No @types/glob-escape package exists
```

?

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+const PLUGIN_NAME: string = 'CopyFilesPlugin';
+const HEFT_STAGE_TAP: TapOptions<'promise'> = {
+  name: PLUGIN_NAME,
+  stage: Number.MAX_SAFE_INTEGER / 2 // This should give us some certainty that this will run after other plugins
+};
+
+const MAX_PARALLELISM: number = 100;
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:37]: May be useful to stick this in https://github.com/microsoft/rushstack/blob/master/apps/heft/src/utilities/Constants.ts

File: apps/heft/src/plugins/DeleteGlobsPlugin.ts [hunk 5]
```diff
@@ -29,6 +29,8 @@ const HEFT_STAGE_TAP: TapOptions<'promise'> = {
   stage: Number.MIN_SAFE_INTEGER
 };
 
+const MAX_PARALLELISM: number = 100;
```
Comment (iclanton) [inline apps/heft/src/plugins/DeleteGlobsPlugin.ts:32]: And we can dedupe this.

File: apps/heft/src/schemas/heft.schema.json [hunk 6]
```diff
@@ -65,6 +65,83 @@
...
+                "properties": {
+                  "actionKind": {
+                    "type": "string",
+                    "enum": ["copyFiles"]
+                  },
+
+                  "heftEvent": {
+                    "type": "string",
+                    "enum": ["pre-compile", "compile", "bundle", "post-build"]
+                  },
```
Comment (iclanton) [inline apps/heft/src/schemas/heft.schema.json:81]: It would be useful to extract this into a shared resource. Take a look at [`command-line.schema.json`](https://github.com/microsoft/rushstack/blob/master/apps/rush-lib/src/schemas/command-line.schema.json) for examples.

File: common/changes/@rushstack/heft/danade-copy-plugin_2020-11-02-23-47.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft",
+      "comment": "Add new default Heft action \"copyGlobs\" to copy or hardlink files during specified Heft events",
```
Comment (iclanton) [inline common/changes/@rushstack/heft/danade-copy-plugin_2020-11-02-23-47.json:5]:
```suggestion
      "comment": "Add new built-in Heft action \"copyGlobs\" to copy or hardlink files during specified Heft events",
```

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  logger: ScopedLogger;
+}
+
+export class CopyFilesPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    const logger: ScopedLogger = heftSession.requestScopedLogger('copy-files');
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:57]:
```suggestion
    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
      const logger: ScopedLogger = heftSession.requestScopedLogger('copy-files');
      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
```

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+    const [copyCount, hardlinkCount] = await this.copyFilesAsync(copyDescriptors);
+    const duration: number = performance.now() - startTime;
+    logger.terminal.writeLine(
+      `Copied ${copyCount} file${copyCount === 1 ? '' : 's'} and linked ${hardlinkCount} ` +
+        `file${hardlinkCount === 1 ? '' : 's'} in ${Math.round(duration)}ms`
+    );
+  }
+
+  protected async copyFilesAsync(copyDescriptors: ICopyFileDescriptor[]): Promise<[number, number]> {
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:127]: Can we return an object with named properties?

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+              throw e;
+            }
+            // Since the file doesn't exist, the parent folder may also not exist
+            await FileSystem.ensureFolderAsync(path.dirname(copyDescriptor.destinationFilePath));
+          }
+
+          await FileSystem.createHardLinkAsync({
+            linkTargetPath: copyDescriptor.sourceFilePath,
+            newLinkPath: copyDescriptor.destinationFilePath
+          });
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:153]: This feels fairly similar to [this code](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts#L154-L172)

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+          hardlinkCount++;
+        } else {
+          // If it's a copy, simply call the copy function
+          await FileSystem.copyFileAsync({
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:157]: This ensures the folder exists by default?

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+const HEFT_STAGE_TAP: TapOptions<'promise'> = {
+  name: PLUGIN_NAME,
+  stage: Number.MAX_SAFE_INTEGER / 2 // This should give us some certainty that this will run after other plugins
+};
+
+const MAX_PARALLELISM: number = 100;
+
+export interface ICopyFileDescriptor {
+  sourceFilePath: string;
+  destinationFilePath: string;
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:41]: It'd be more efficient for copy (i.e. - not hardlink) operations to read the file once and write it `n` times instead of repeatedly calling `FileSystem.copyFileAsync`, which will read the source file from the filesystem for every destination.

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+      // Glob extensions with a specific glob to increase perf
+      let sourceFileRelativePaths: Set<string>;
+      if (copyConfiguration.fileExtensions?.length) {
+        const escapedExtensions: string[] = globEscape(copyConfiguration.fileExtensions);
+        const pattern: string = `**/*+(${escapedExtensions.join('|')})`;
+        sourceFileRelativePaths = await this._expandGlobPatternAsync(
+          resolvedSourceFolderPath,
+          pattern,
+          copyConfiguration.excludeGlobs
+        );
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:191]: I wonder if this would be faster if we crawled the filesystem explicitly and looked for files matching one of the extensions instead of constructing and evaluating a glob.

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+      } else {
+        sourceFileRelativePaths = new Set<string>();
+      }
+
+      // Now include the other glob as well
+      for (const include of copyConfiguration.includeGlobs || []) {
+        const explicitlyIncludedPaths: Set<string> = await this._expandGlobPatternAsync(
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:198]: It'd be cool if we could somehow parse these glob patterns to compiled regexes or something and then try to match those against an explicit crawling of the filesystem.

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+
+      // Dedupe and throw if a double-write is detected
+      for (const destinationFolderRelativePath of copyConfiguration.destinationFolders) {
+        for (const sourceFileRelativePath of sourceFileRelativePaths) {
+          // Only include the relative path from the sourceFolder if flatten is false
+          const resolvedSourceFilePath: string = path.join(resolvedSourceFolderPath, sourceFileRelativePath);
+          const resolvedDestinationFilePath: string = path.resolve(
+            buildFolder,
+            destinationFolderRelativePath,
+            copyConfiguration.flatten ? '.' : path.dirname(sourceFileRelativePath),
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:217]: I'm starting to wonder if `flatten` is really a useful option.

File: apps/heft/src/plugins/CopyStaticAssetsPlugin.ts [hunk 16]
```diff
@@ -103,147 +64,72 @@ export class CopyStaticAssetsPlugin implements IHeftPlugin {
...
-    return copyCount;
   }
 
-  private async _runCopyAsync(options: ICopyStaticAssetsOptions): Promise<void> {
-    const { logger, buildFolder, copyStaticAssetsConfiguration, watchMode } = options;
-
-    if (!copyStaticAssetsConfiguration.sourceFolderName) {
-      return;
-    }
+  private async _runWatchAsync(options: ICopyStaticAssetsOptions): Promise<void> {
```
Comment (iclanton) [inline apps/heft/src/plugins/CopyStaticAssetsPlugin.ts:96]: How hard would it be to provide this functionality for the base copy plugin?

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 11 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:157]: Yes it does :)

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 13 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:191]: Switched to fast-glob. We talked offline, but under the covers this uses `micromatch` which is what I was planning on using anyway, alongside a manual crawl of the directory. We're also able to do a single call with multiple different glob patterns in one go, avoiding enumerating the filesystem multiple times when running patterns in the same folder. Ideally this should have the same perf as a custom solution, save for the additional calls we would have to make when targeting other `sourceFolder`s

Another thought here would be that we could combine all globs across all copy configurations, while appending the individual `sourceFolder` path to the glob, though I'm unsure if this would cause a perf hit, ie:
```json
{
  "sourceFolder": "src/subfolder",
  "includeGlobs": [ "**/*.json", "**/*.md" ] // -> "includeGlobs": [ "src/subfolder/**/*.json", "src/subfolder/**/*.md" ]
  ...
}
```

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 14 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:198]: My thought is that if compiled regexes are faster, this would be the approach fast-glob would take internally. Either that, or at the very least it would provide the option to do this. `micromatch`, which `fast-glob` relies on, has the ability to convert a glob into a regex, and I think fast-glob would try to take advantage of it given the speed-focused mentality of the lib. I could very easily be wrong here though.

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 15 above)
Comment (D4N14L) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:217]: I still like it in the context of this being a "general purpose" plugin. Even if it's a feature that not many use, it's not exactly much effort to maintain.

Though, if you still hate it, we can pull it out

File: apps/heft/src/schemas/heft.schema.json (hunk 6 above)
Comment (D4N14L) [reply apps/heft/src/schemas/heft.schema.json:81]: I didn't do this, specifically because we purposely omit "clean" from that list, and according to json-schema spec, you cannot override a `$ref` definition :(

Comment (D4N14L) [reply apps/heft/src/schemas/heft.schema.json:81]: `Any members other than "$ref" in a JSON Reference object SHALL be ignored.`
https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-3

File: libraries/node-core-library/src/FileSystem.ts [hunk 17]
```diff
@@ -911,6 +941,100 @@ export class FileSystem {
...
+   */
+  public static async copyFileToManyAsync(options: IFileSystemCopyFileToManyOptions): Promise<void> {
+    options = {
+      ...COPY_FILE_DEFAULT_OPTIONS,
+      ...options
+    };
+
+    if (FileSystem.getStatistics(options.sourcePath).isDirectory()) {
+      throw new Error(
+        'The specified path refers to a folder; this operation expects a file object:\n' + options.sourcePath
```
Comment (iclanton) [inline libraries/node-core-library/src/FileSystem.ts:962]:
```suggestion
        'The specified path refers to a folder; this operation expects a file path:\n' + options.sourcePath
```

File: libraries/node-core-library/src/FileSystem.ts [hunk 18]
```diff
@@ -911,6 +941,100 @@ export class FileSystem {
...
+                reject(e);
+              }
+            });
+        });
+      };
+
+      const sourceStream: fs.ReadStream = fsx.createReadStream(options.sourcePath);
+      const uniqueDestinationPaths: Set<string> = new Set(options.destinationPaths);
+      const pipePromises: Promise<void>[] = [];
+      for (const destinationPath of uniqueDestinationPaths.values()) {
```
Comment (iclanton) [inline libraries/node-core-library/src/FileSystem.ts:1030]:
```suggestion
      for (const destinationPath of uniqueDestinationPaths) {
```

File: common/config/rush/browser-approved-packages.json [hunk 19]
```diff
@@ -6,6 +6,10 @@
       "name": "@rushstack/heft-web-rig",
       "allowedCategories": [ "libraries", "tests" ]
     },
+    {
+      "name": "fast-glob",
+      "allowedCategories": [ "libraries" ]
+    },
```
Comment (iclanton) [inline common/config/rush/browser-approved-packages.json:12]: Move this to `nonbrowser-approved-packages.json`

File: common/changes/@rushstack/heft/danade-copy-plugin_2020-11-02-23-47.json [hunk 20]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft",
+      "comment": "Add new built-in Heft action \"copyGlobs\" to copy or hardlink files during specified Heft events",
```
Comment (iclanton) [inline common/changes/@rushstack/heft/danade-copy-plugin_2020-11-02-23-47.json:5]:
```suggestion
      "comment": "Add new built-in Heft action \"copyFiles\" to copy or hardlink files during specified Heft events",
```

File: common/config/rush/browser-approved-packages.json [hunk 21]
```diff
@@ -2,6 +2,10 @@
 {
   "$schema": "https://developer.microsoft.com/json-schemas/rush/v5/approved-packages.schema.json",
   "packages": [
+    {
+      "name": "fast-glob",
```
Comment (octogonz) [inline common/config/rush/browser-approved-packages.json:6]: Seems like this should go in `nonbrowser-approved-packages.json`

File: common/reviews/api/node-core-library.api.md [hunk 22]
```diff
@@ -298,8 +299,16 @@ export interface IFileSystemCopyFilesOptions extends IFileSystemCopyFilesAsyncOp
     filter?: FileSystemCopyFilesFilter;
 }
 
+// @public
+export interface IFileSystemCopyFileToManyOptions extends Omit<IFileSystemCopyFileOptions, 'destinationPath'> {
```
Comment (octogonz) [inline common/reviews/api/node-core-library.api.md:303]: Can we design this interface without `Omit<>`?  Algebraic types are difficult to represent on an API website.

File: libraries/node-core-library/src/FileSystem.ts [hunk 23]
```diff
@@ -911,6 +933,100 @@ export class FileSystem {
...
+              streamToDestroy.destroy();
+            }
+            reject(e);
+          });
+          sourceStream
+            .pipe(destinationStream)
+            .on('close', () => {
+              resolve();
+            })
+            .on('error', async (e: Error) => {
```
Comment (octogonz) [inline libraries/node-core-library/src/FileSystem.ts:990]: This seems incorrect.  Do Node.js streams support promises?  If not, then if the `Promise` rejects there will be no catch handler, and the process will get terminated.

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  private async _getCopyFileDescriptorsAsync(
+    buildFolder: string,
+    copyConfigurations: IExtendedSharedCopyConfiguration[]
+  ): Promise<ICopyFileDescriptor[]> {
+    // Create a map to deduplicate and prevent double-writes
+    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+    // And a map to contain the actual results
+    const sourceCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:195]: For `destinationCopyDescriptors` and `sourceCopyDescriptors` document what the `Map` key is.  Absolute path to a source file?

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+    return Array.from(sourceCopyDescriptors.values());
+  }
+
+  private _getIncludedGlobPatterns(copyConfiguration: IExtendedSharedCopyConfiguration): string[] {
+    // Glob extensions with a specific glob to increase perf
+    const patternsToGlob: Set<string> = new Set<string>();
+    for (const fileExtension of copyConfiguration.fileExtensions || []) {
+      const escapedExtension: string = glob.escapePath(fileExtension);
+      patternsToGlob.add(`**/*${escapedExtension}`);
+    }
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:269]: Would this:
```
**/*.{png,gif,jpg,jpeg,svg}
```
...be significantly more efficient than this?
```
**/*.png
**/*.gif
**/*.jpg
**/*.jpeg
**/*.svg
```
If so it might be worth optimizing them, since this should be the 99% most common case.

(Ideally we could later optimize it even further and skip invoking the glob engine entirely, if the pattern only specifies file extensions.)

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+              });
+            }
+          );
+          await Promise.all(hardlinkPromises);
+
+          linkedFileCount++;
+        } else {
+          // If it's a copy, we will call the copy function
+          if (copyDescriptor.destinationFilePaths.length === 1) {
+            await FileSystem.copyFileAsync({
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:164]: What ensures that the target folder exists?

File: apps/heft/src/plugins/CopyFilesPlugin.ts [hunk 27]
```diff
@@ -0,0 +1,336 @@
...
+
+const PLUGIN_NAME: string = 'CopyFilesPlugin';
+const HEFT_STAGE_TAP: TapOptions<'promise'> = {
+  name: PLUGIN_NAME,
+  stage: Number.MAX_SAFE_INTEGER / 2 // This should give us some certainty that this will run after other plugins
+};
+
+interface ICopyFileDescriptor {
+  sourceFilePath: string;
+  destinationFilePaths: string[];
```
Comment (octogonz) [inline apps/heft/src/plugins/CopyFilesPlugin.ts:39]: Would it be much work to create a test project that illustrates some common cases:
- Copying one file
- Copying one file to multiple target folders
- Copying several files in a nested folder tree
- Copying a file into a folder that does not already exist

**Review by octogonz — APPROVED** [review]

File: apps/heft/src/schemas/heft.schema.json [hunk 28]
```diff
@@ -65,6 +65,83 @@
...
+                    "type": "array",
+                    "description": "An array of copy operations to run perform during the specified Heft event.",
+                    "items": {
+                      "type": "object",
+                      "required": ["sourceFolder", "destinationFolders"],
+                      "properties": {
+                        "sourceFolder": {
+                          "type": "string",
+                          "description": "The folder from which files should be copied.",
+                          "pattern": "[^\\\\]"
```
Comment (octogonz) [inline apps/heft/src/schemas/heft.schema.json:93]: Is it okay for Heft to read from or write to files outside the project folder?  If not, should we restrict that somehow?

**Comment by octogonz** [comment]
I've approved this PR.  I don't want to block it.  But I'm still not 100% sure about the perf implications of our design.

Ideally the config file should encourage developers NOT to write globs that do an `O(n)` scan of every subdirectory.  Optimizing our globs is also good 👍 👍 , but NOT using globs in the first place will probably always be faster.  Our design shouldn't prevent globs, just make it clear when they are inefficient.

The idea I originally advocated seems to end up like this, which is complicated and awkward:
- `filesToInclude`
- `foldersToInclude`
- `foldersToExclude`
- `globsToInclude`
- `globsToExclude`

Maybe a simpler idea would be:
- `globsToInclude`
- `globsToExclude`
- `allowSlowGlobs`

In other words, what if our engine detected inefficient globs (e.g. `**`), such that the person had to explicitly set `allowSlowGlobs=true` to say *"yes, I realize my approach is going to be inefficient."*  

The `fast-glob` engine distinguishes [static-vs-dynamic patterns](https://www.npmjs.com/package/fast-glob#what-is-a-static-or-dynamic-pattern), which maybe is the fast-vs-slow concept. 

Just something to think about... again, I don't want to hold up this PR.

File: apps/heft/src/schemas/heft.schema.json (hunk 28 above)
Comment (iclanton) [reply apps/heft/src/schemas/heft.schema.json:93]: I think that's fine for now. We may want to come up with a stricter set of rules at some point, though.

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 26 above)
Comment (iclanton) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:164]: `fsx.copy` does that internally.

File: apps/heft/src/plugins/CopyFilesPlugin.ts (hunk 27 above)
Comment (iclanton) [reply apps/heft/src/plugins/CopyFilesPlugin.ts:39]: Added a test project.

**Comment by octogonz** [comment]
The JSON schema changes from this PR have been published.

File: libraries/node-core-library/src/FileSystem.ts [hunk 29]
```diff
@@ -260,6 +277,11 @@ export interface IFileSystemCreateLinkOptions {
    * The new path for the new symlink link to be created.
    */
   newLinkPath: string;
+
+  /**
+   * Specifies what to do if the target object already exists. Defaults to `AlreadyExistsBehavior.Error`.
+   */
+  alreadyExistsBehavior?: AlreadyExistsBehavior;
```
Comment (octogonz) [inline libraries/node-core-library/src/FileSystem.ts:284]: @D4N14L  I noticed today that this new option applies to lots of APIs like `createSymbolicLinkFolder()` and `createSymbolicLinkJunction()`, but it is only implemented for `createHardLink()`

Comment (D4N14L) [reply libraries/node-core-library/src/FileSystem.ts:284]: I can make a quick PR to update.
